src/MongoDB.Driver/ClientSessionHandle.cs

```diff
@@ -144,5 +244,162 @@ private TransactionOptions GetEffectiveTransactionOptions(TransactionOptions tra
 
             return new TransactionOptions(readConcern, readPreference, writeConcern);
         }
+
+        private WithTransactionStepResult<TResult> WithTransactionCallbackProcessing<TResult>(Func<IClientSessionHandle, TResult> callback, DateTime startTime, CancellationToken cancellationToken)
+        {
+            TResult result = default;
+            try
+            {
+                result = callback(this);
+                return new WithTransactionStepResult<TResult>(result, false);
+            }
+            catch (Exception ex)
+            {
+                if (IsInTransaction)
+                {
+                    AbortTransaction(cancellationToken);
+                }
+
+                if (ex is MongoException mongoException)
+                {
+                    var now = _dateTimeProvider.Now;
+                    if (mongoException.HasErrorLabel(TransientTransactionErrorLabel) && !IsTimeout(startTime, now))
+                    {
+                        return new WithTransactionStepResult<TResult>(result, true);
+                    }
+
+                    if (mongoException.HasErrorLabel(UnknownTransactionCommitResultLabel))
+                    {
+                        CommitTransaction(cancellationToken);
+                    }
+                }
+
+                throw;
+            }
+        }
+
+        private async Task<WithTransactionStepResult<TResult>> WithTransactionCallbackProcessingAsync<TResult>(Func<IClientSessionHandle, TResult> callback, DateTime startTime, CancellationToken cancellationToken)
```

Comment:
Something about this seems wrong to me, but it may just be my lack of understanding of async in C#.  

Does this require, or should it, that the callback itself uses the async API?  Does it work whether it does or it doesn't?

Reply:
Good question. 
To call await methods in a callback, the callback definition should be like:
      
    public async Task<TResult> WithTransactionAsync<TResult>(Func<IClientSessionHandle, Task<TResult>> callback, TransactionOptions transactionOptions = null, CancellationToken cancellationToken = default(CancellationToken))

with `Task` in the Func response part. In this case we will be able to call await methods in callback in this way:

    var result = subject.WithTransactionAsync<bool>(async (handle) =>
    {
          return await SomeAsyncMethod();
    })

I will change the method definition.
But how do you think, will users expect method overload with not `Task` `Func` (like it's implemented right now), I mean whether they wanted to call sync callbacks with async `CommitTransaction` / `AbortTransaction` or not?
Calling will be like this:

    var result = subject.WithTransactionAsync<bool>((handle) =>
    {
          return SomeSyncMethod();
    })

Reply:
Applications should either be 100% sync or 100% async.

Mixing sync callback with WithTransactionAsync is an anti-pattern.

Reply:
Done